rust/src/runner/context.rs

```diff
@@ -286,3 +275,243 @@ impl Drop for TestContext {
         }
     }
 }
+
+/// Allows to create a file using builder pattern.
+#[derive(Debug)]
+pub struct FileBuilder {
+    file_type: FileType,
+    path: PathBuf,
+    random_name: bool,
+    mode: Mode,
+}
+
+impl FileBuilder {
+    /// Create a file builder.
+    pub fn new<P: AsRef<Path>>(file_type: FileType, base_path: &P) -> Self {
+        Self {
+            file_type: file_type.clone(),
+            path: base_path.as_ref().to_path_buf(),
+            random_name: true,
+            mode: match file_type {
+                FileType::Dir => Mode::from_bits_truncate(0o755),
+                _ => Mode::from_bits_truncate(0o644),
+            },
+        }
+    }
+
+    /// Create the file according to the provided information.
+    pub fn create(self) -> nix::Result<PathBuf> {
+        let mode = self.mode;
+        let mut path = self.path;
+        if self.random_name {
+            path = path.join(
+                thread_rng()
+                    .sample_iter(&Alphanumeric)
+                    .take(NUM_RAND_CHARS)
+                    .map(char::from)
+                    .collect::<String>(),
+            );
+        }
+
+        match self.file_type {
+            FileType::Regular => open(&path, OFlag::O_CREAT, mode).and_then(close),
+            FileType::Dir => mkdir(&path, mode),
+            FileType::Fifo => mkfifo(&path, mode),
+            FileType::Block => mknod(&path, SFlag::S_IFBLK, mode, 0),
+            FileType::Char => mknod(&path, SFlag::S_IFCHR, mode, 0),
+            FileType::Socket => {
+                let fd = socket(
+                    nix::sys::socket::AddressFamily::Unix,
+                    nix::sys::socket::SockType::Stream,
+                    SockFlag::empty(),
+                    None,
+                )?;
+                let sockaddr = UnixAddr::new(&path)?;
+                bind(fd, &sockaddr)
+            }
+            FileType::Symlink(target) => symlink(
+                target.as_deref().unwrap_or_else(|| Path::new("test")),
+                &path,
+            )
+            .map_err(|e| nix::Error::try_from(e).unwrap_or(nix::errno::Errno::UnknownErrno)),
+        }?;
+
+        Ok(path)
+    }
+
+    /// Create the file according to the provided information and open it.
+    pub fn open(self, oflags: OFlag) -> nix::Result<(PathBuf, RawFd)> {
```

Comment:
I think there are some tests that create a file with mode 0o444 and test that they can still write to it.  Those tests depend on the file being opened by the same syscall that created it.  They'll fail if you use this method, which closes the file and then reopens it.

Reply:
Thanks, I fixed it!